File: src/main/java/io/vproxy/vfx/manager/audio/AudioManager.java
package io.vproxy.vfx.manager.audio;

import io.vproxy.base.util.LogType;
import io.vproxy.base.util.Logger;
import javafx.scene.media.AudioClip;

import java.io.IOException;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ConcurrentHashMap;

public class AudioManager {
    private static final AudioManager instance = new AudioManager();

    public static AudioManager get() {
        return instance;
    }

    private AudioManager() {
    }

    private final Map<String, AudioClip> map = new ConcurrentHashMap<>();
    private final WeakHashMap<String, AudioClip> weakMap = new WeakHashMap<>();

    public AudioClip loadAudio(String path) {
        try {
            return loadAudio(path, false);
        } catch (Exception ignore) {
            return null;
        }
    }

    @SuppressWarnings("RedundantThrows")
    public AudioClip loadAudio(String path, boolean throwException) throws Exception {
        if (path.startsWith("/")) {
            path = path.substring(1);
        }
        var audio = map.get(path);
        if (audio == null) {
            audio = weakMap.get(path);
        }
        if (audio != null) {
            assert Logger.lowLevelDebug("using cached audio: " + path);
            return audio;
        }
        try {
            var res = getClass().getClassLoader().getResource(path);
            if (res == null) {
                Logger.error(LogType.FILE_ERROR, "unable to find resource for audio " + path);
                if (throwException) {
                    throw new IOException("cannot find audio " + path);
                }
                return null;
            }
            audio = new AudioClip(res.toExternalForm());
        } catch (Exception e) {
            Logger.error(LogType.FILE_ERROR, "failed loading audio " + path, e);
            if (throwException) {
                throw e;
            }
            return null;
        }
        map.put(path, audio);
        assert Logger.lowLevelDebug("new audio loaded: " + path);
        return audio;
    }

    public AudioClip weakRefAudio(String path) {
        var audio = map.remove(path);
        if (audio == null) {
            return null;
        }
        weakMap.put(path, audio);
        return audio;
    }

    public void removeAudio(String path) {
        if (!map.containsKey(path) && !weakMap.containsKey(path)) {
            return;
        }
        map.remove(path);
        weakMap.remove(path);
        assert Logger.lowLevelDebug("audio removed: " + path);
    }
}

File: src-ts/src/manager/audio/AudioManager.ts
import { Logger } from '../../vproxy-base/Logger.js';

export type AudioClip = HTMLAudioElement;

function toAudioAssetUrl(path: string): string {
  // Java strips leading '/'; we map "io/vproxy/..." to "./res/audio/...".
  // vfx does not bundle audio resources, so this is a placeholder mapping.
  if (path.startsWith('/')) path = path.substring(1);
  const idx = path.indexOf('io/vproxy/vfx/res/audio/');
  if (idx >= 0) {
    return './res/audio/' + path.substring(idx + 'io/vproxy/vfx/res/audio/'.length);
  }
  if (path.startsWith('io/vproxy/vfx/res/')) {
    return './res/' + path.substring('io/vproxy/vfx/res/'.length);
  }
  return './' + path;
}

export class AudioManager {
  static readonly instance = new AudioManager();

  static get(): AudioManager { return AudioManager.instance; }

  private readonly map = new Map<string, AudioClip>();
  private readonly weakMap = new Map<string, WeakRef<AudioClip>>();

  private constructor() {}

  loadAudio(path: string): AudioClip | null {
    if (path.startsWith('/')) path = path.substring(1);
    const cached = this.map.get(path);
    if (cached) return cached;
    const weak = this.weakMap.get(path)?.deref();
    if (weak) return weak;

    try {
      const url = toAudioAssetUrl(path);
      const clip = new Audio(url);
      clip.preload = 'auto';
      this.map.set(path, clip);
      return clip;
    } catch (e) {
      Logger.error('FILE_ERROR', `failed loading audio ${path}`, e as Error);
      return null;
    }
  }

  weakRefAudio(path: string): AudioClip | null {
    const audio = this.map.get(path);
    if (!audio) return null;
    this.map.delete(path);
    this.weakMap.set(path, new WeakRef(audio));
    return audio;
  }

  removeAudio(path: string): void {
    if (path.startsWith('/')) path = path.substring(1);
    if (!this.map.has(path) && !this.weakMap.has(path)) return;
    this.map.delete(path);
    this.weakMap.delete(path);
  }
}
